scripts/codegen/devicetree.py

```diff
@@ -0,0 +1,243 @@
+                                                 default="<unset>"):
+        controller_prefix = self.device_tree_controller_property(
+                                device_type_name, device_index, property_name, None)
+        if controller_prefix is None:
+            if default == "<unset>":
+                controller_prefix = self.device_tree_controller_prefix(device_type_name, device_index)
+                default = \
+                "Device tree property {}_{} not defined.".format(controller_prefix, property_name)
+            return default
+        property_name = "{}_{}".format(controller_prefix, property_name_indirect)
+        return self._device_tree_property(property_name, default)
+
+    ##
+    # @brief Check compatibility to device given by type and idx.
+    #
+    # The @p compatible parameter is checked against the compatible property of
+    # the device given by @p device_type_name and @p device_index.
+    #
+    # @param device_type_name Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
+    # @param device_index Index of device
+    # @param compatible driver compatibility string (e.g. st,stm32-spi-fifo)
+    # @return 1 if compatible, 0 otherwise
+    #
+    def device_tree_controller_compatible_x(self, device_type_name,
+                                            device_index, compatible,
+                                            default="<unset>"):
+        compatible_count = self.device_tree_controller_property(device_type_name, device_index,
+                                                                'COMPATIBLE_COUNT', None)
+        if compatible_count is None or int(compatible_count) == 0:
+            return 0
+        compatible_count = int(compatible_count)
+        # sanitize for label usage
+        compatible = self._device_tree_label(compatible)
+        compatible_id_property_name = "COMPATIBLE_ID_{}".format(compatible)
+        compatible_id = self._device_tree_property(compatible_id_property_name, None)
+        if compatible_id is None:
+            if default == "<unset>":
+                default = \
+                "Device tree property COMPATIBLE_ID_{} not defined.".format(compatible)
+            return default
+        for x in range(0, compatible_count):
+            controller_compatibe_id_property_label = "COMPATIBLE_{}_ID".format(x)
+            controller_compatibe_id = self.device_tree_controller_property(
+                device_type_name, device_index, controller_compatibe_id_property_label, None)
+            if controller_compatibe_id == compatible_id:
+                return 1
+        return 0
+
+    ##
+    # @brief Check compatibility to at least one activated device of given type.
+    #
+    # The @p compatible parameter is checked against the compatible property of
+    # all activated devices of given @p type.
+    #
+    # @param device_type_name Type of device controller (e.g. 'SPI', 'GPIO', 'PIN', ...)
```

Comment:
I think this is an abuse of device tree specification.
Not every node could be named as controller. spi-controller is not documented anywhere and htere is one example of i2c-controller which is documented in Linux as a controller of i2c nodes.
I think you should find another solution to identify devices, and not assume they all are controller.